src/inputfetcher.h

```diff
@@ -0,0 +1,197 @@
+                // If an input spends an outpoint from earlier in the block,
+                // it won't be in the cache yet but it also won't be in the db either.
+                if (m_txids.contains(input.outpoint.hash)) {
+                    input.status.store(Input::Status::SKIPPED, std::memory_order_relaxed);
+                    continue;
+                }
+                try {
+                    if (auto coin{m_cache->GetPossiblySpentCoinFromCache(input.outpoint)}) {
+                        input.coin = std::move(*coin);
+                        if (!input.coin.IsSpent()) [[likely]] { // Coin from cache could be spent
+                            // We need release here, so setting coin 2 lines above happens before the main thread loads.
+                            input.status.store(Input::Status::READY, std::memory_order_release);
+                            continue;
+                        }
+                    } else if (auto coin{m_db->GetCoin(input.outpoint)}) {
+                        input.coin = std::move(*coin); // Coin from db cannot be spent, it does not store spent coins
+                        // We need release here, so setting coin in the previous line happens before the main thread loads.
+                        input.status.store(Input::Status::READY, std::memory_order_release);
+                        continue;
+                    }
+                } catch (const std::runtime_error& e) {
+                    LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Warning, "InputFetcher failed to fetch input: %s.\n", e.what());
+                }
+                // Input missing or spent. This block will fail validation.
+                // Skip remaining inputs.
+                m_input_head.store(m_inputs.size(), std::memory_order_relaxed);
+                input.status.store(Input::Status::FAILED, std::memory_order_relaxed);
+            }
+            m_barrier.arrive_and_wait();
+        }
+    }
+
+public:
+    explicit InputFetcher(size_t worker_thread_count) noexcept
+        : m_barrier{static_cast<int32_t>(worker_thread_count + 1)}
+    {
+        for (size_t n{0}; n < worker_thread_count; ++n) {
+            m_worker_threads.emplace_back([this, n]() {
+                util::ThreadRename(strprintf("inputfetch.%i", n));
+                WorkLoop();
+            });
+        }
+    }
+
+    //! Fetch all block inputs from cache or db, and insert into temp_cache.
+    void FetchInputs(CCoinsViewCache& temp_cache, const CCoinsViewCache& cache, const CCoinsView& db, const CBlock& block) noexcept
+    {
+        if (block.vtx.size() <= 1 || m_worker_threads.size() == 0) {
+            return;
+        }
+
+        // Loop through the inputs of the block and set them in the queue.
+        // Construct the set of txids to filter, and count the outputs to reserve for temp_cache.
+        auto outputs_count{block.vtx[0]->vout.size()};
+        for (size_t i{1}; i < block.vtx.size(); ++i) {
```

Comment:
we're also just adding stuff to the `m_inputs` without any reservation, the first few times we will have some needless copying - we could likely alleviate that by doing a rough reservation
```C++
        m_txids.reserve(block.vtx.size());
        m_inputs.reserve(2 * block.vtx.size()); // rough guess
```

Reply:
Did it anyways.